### PR TITLE
Issue 1714

### PR DESCRIPTION
**Pull request by koheiw**
I mistakenly based this branch on `issue-1375`, so I can wait until the branch is merged.

This PR basically 

- Adds `y` to compute similarity between non-intersecting documents/features, and deprecate `selection`
- Changes the underlying Matrix to "dgeMatrix" and "dspMatrix" when `min_simil` is not used to make the resulting objects compact. 

When `min_simil` is used, the objects remains "dgTMatrix" as their sizes are not the main concern in this case

### CODE REVIEW

**Comment by koheiw** [comment]
https://github.com/quanteda/quanteda/blob/bca42a2dfa88f2fa156e2fb21d1c6dd0d6e094e9/R/textstat_simil.R#L497-L499
is not really correct, because `proxyC::simil()` returns explicit zeros.

File: R/textstat_simil.R [hunk 1]
```diff
@@ -16,100 +16,93 @@
...
-#' @rdname textstat_simildist-class
-setClass("textstat_simil_sel", contains = c("textstat_simildist", "dgTMatrix"),
-         slots = c(selection = "character"))
+#' @rdname textstat_proxy-class
+setClass("textstat_simil", contains = c("textstat_proxy", "dgeMatrix"))
 
-#' @rdname textstat_simildist-class
-setClass("textstat_simil_sparse", contains = c("textstat_simildist", "dsCMatrix"),
+#' @rdname textstat_proxy-class
+setClass("textstat_simil_sparse", contains = c("textstat_proxy", "dgTMatrix"),
```
[the remark this answers is not in the document]
Comment (koheiw) [reply R/textstat_simil.R]: Fixed

File: R/textstat_simil.R [hunk 2]
```diff
@@ -210,54 +207,48 @@ textstat_simil.dfm <- function(x, selection = NULL,
...
-    
-    if (min_simil == 0) {
-        if (is.null(selection))
-            return(new("textstat_simil", as(temp, "dsTMatrix"), 
+    temp <- textstat_proxy(x, y, margin, method,
+                           min_proxy = min_simil, use_na = TRUE)
+
+    if (is.null(min_simil)) {
+        if (identical(x, y)) {
+            return(new("textstat_simil_symm", as(temp, "dsyMatrix"),
```
[the remark this answers is not in the document]
Comment (koheiw) [reply R/textstat_simil.R]: Fixed

File: R/textstat_simil.R [hunk 3]
```diff
@@ -16,100 +16,93 @@
...
+#' @rdname textstat_proxy-class
 #' @slot selection target units, if any
-setClass("textstat_dist_sel", contains = c("textstat_simildist", "dgTMatrix"),
-         slots = c(selection = "character"))
+setClass("textstat_dist", contains = c("textstat_proxy", "dgeMatrix"))
 
-#' @rdname textstat_simildist-class
-setClass("textstat_simil", contains = c("textstat_simildist", "dsTMatrix"))
+#' @rdname textstat_proxy-class
+setClass("textstat_simil_symm", contains = c("textstat_proxy", "dspMatrix"))
```
[the remark this answers is not in the document]
Comment (koheiw) [reply R/textstat_simil.R]: Restored

File: R/textstat_simil.R [hunk 4]
```diff
@@ -16,100 +16,93 @@
...
-setClass("textstat_simildist", contains = "dsparseMatrix",
+setClass("textstat_proxy", contains = "Matrix",
          slots = c(method = "character",
                    margin = "character",
                    type = "character"))
 
-#' @rdname textstat_simildist-class
-setClass("textstat_dist", contains = c("textstat_simildist", "dsTMatrix"))
+#' @rdname textstat_proxy-class
+setClass("textstat_dist_symm", contains = c("textstat_proxy", "dspMatrix"))
```
[the remark this answers is not in the document]
Comment (koheiw) [reply R/textstat_simil.R]: dsp is the smallest of all.

File: R/textstat_simil.R [hunk 5]
```diff
@@ -16,100 +16,93 @@
 #'   \code{"features"} for word/term features.
 #' @slot type either \code{"textstat_simil"} or \code{"textstat_dist"}
 #' @seealso \code{\link{textstat_simil}}
-setClass("textstat_simildist", contains = "dsparseMatrix",
```
[the remark this answers is not in the document]
Comment (koheiw) [reply R/textstat_simil.R]: Objects are dense matrices most of the cases.

**Comment by kbenoit** [comment]
@koheiw I see you have deprecated `selection`, which is fine, so I added a note about this to the Rd.  Should we consider removing it from the signature but trapping it in `...`? Or is the note enough to discourage its use?

Also can you fix the warning here? After that, I'm happy for this to be merged.

``` r
library("quanteda")
## Package version: 1.4.5
dfmat <- dfm(corpus_subset(data_corpus_inaugural, Year > 2000),
  remove_punct = TRUE, remove = stopwords("english")
)
tstat5 <- textstat_dist(dfmat, dfmat[c("2009-Obama", "2013-Obama"), ], margin = "documents")
as.list(tstat5)
## Warning in proxy2triplet(x, upper = TRUE): upper = TRUE has no effect when
## columns have been selected
## $`2009-Obama`
##  2005-Bush 2017-Trump  2001-Bush 2013-Obama 
##   60.84406   54.91812   49.94997   47.98958 
## 
## $`2013-Obama`
##  2005-Bush 2017-Trump  2001-Bush 2009-Obama 
##   56.85948   53.73081   48.31149   47.98958
```

**Comment by koheiw** [comment]
I think it is better to hide the argument but not necessary.

**Pull request merged**